File: haskell/glue/src/Glue/Lib/List.hs
module Glue.Lib.List where

import Glue.Eval (Eval)
import Glue.Lib.List.Append (append)
import Glue.Lib.List.Butlast (butlast)
import Glue.Lib.List.Concat qualified as Concat
import Glue.Lib.List.Drop qualified as Drop
import Glue.Lib.List.Filter qualified as Filter
import Glue.Lib.List.Find qualified as Find
import Glue.Lib.List.Flatten qualified as Flatten
import Glue.Lib.List.Foldl qualified as Foldl
import Glue.Lib.List.Foldr qualified as Foldr
import Glue.Lib.List.Head qualified as Head
import Glue.Lib.List.Last qualified as Last
import Glue.Lib.List.Length qualified as Length
import Glue.Lib.List.Map qualified as Map
import Glue.Lib.List.Member (member)
import Glue.Lib.List.Partition qualified as Partition
import Glue.Lib.List.Position qualified as Position
import Glue.Lib.List.Prepend (prepend)
import Glue.Lib.List.Remove (remove)
import Glue.Lib.List.Reverse qualified as Reverse
import Glue.Lib.List.Sort qualified as Sort
import Glue.Lib.List.Tail qualified as Tail
import Glue.Lib.List.Take qualified as Take
import Glue.Lib.List.Zip qualified as Zip
import Glue.Module (ModuleInfo, nativeModule)

list :: ModuleInfo Eval
list =
    nativeModule
        "ffi.list"
        [ ("append", append)
        , ("but-last", butlast)
        , ("concat", Concat.concat)
        , ("head", Head.head)
        , ("drop", Drop.drop)
        , ("filter", Filter.filter)
        , ("find", Find.find)
        , ("flatten", Flatten.flatten)
        , ("foldl", Foldl.foldl)
        , ("foldr", Foldr.foldr)
        , ("last", Last.last)
        , ("length", Length.length)
        , ("map", Map.map)
        , ("member", member)
        , ("partition", Partition.partition)
        , ("prepend", prepend)
        , ("position", Position.position)
        , ("remove", remove)
        , ("reverse", Reverse.reverse)
        , ("sort", Sort.sort)
        , ("tail", Tail.tail)
        , ("take", Take.take)
        , ("zip", Zip.zip)
        ]

File: dart/glue/lib/src/lib/list.dart
import 'package:glue/src/lib/list/append.dart';
import 'package:glue/src/lib/list/butlast.dart';
import 'package:glue/src/lib/list/concat.dart';
import 'package:glue/src/lib/list/drop.dart';
import 'package:glue/src/lib/list/filter.dart';
import 'package:glue/src/lib/list/find.dart';
import 'package:glue/src/lib/list/flatten.dart';
import 'package:glue/src/lib/list/foldl.dart';
import 'package:glue/src/lib/list/foldr.dart';
import 'package:glue/src/lib/list/head.dart';
import 'package:glue/src/lib/list/last.dart';
import 'package:glue/src/lib/list/length.dart';
import 'package:glue/src/lib/list/map.dart';
import 'package:glue/src/lib/list/member.dart';
import 'package:glue/src/lib/list/partition.dart';
import 'package:glue/src/lib/list/position.dart';
import 'package:glue/src/lib/list/prepend.dart';
import 'package:glue/src/lib/list/remove.dart';
import 'package:glue/src/lib/list/reverse.dart';
import 'package:glue/src/lib/list/sort.dart';
import 'package:glue/src/lib/list/tail.dart';
import 'package:glue/src/lib/list/take.dart';
import 'package:glue/src/lib/list/zip.dart';
import 'package:glue/src/module.dart';

/// List module - list manipulation functions
/// Mirrors Haskell Glue.Lib.List exactly

/// The list module containing all list functions
/// Mirrors Haskell Glue.Lib.List.list exactly
final ModuleInfo listModule = nativeModule('ffi.list', [
  // Core list operations
  ('append', append),
  ('butlast', butlast),
  ('concat', concat),
  ('drop', drop),
  ('filter', filter),
  ('find', find),
  ('flatten', flatten),
  ('foldl', foldl),
  ('foldr', foldr),
  ('head', head),
  ('last', last),
  ('length', length),
  ('map', map),
  ('member', member),
  ('partition', partition),
  ('position', position),
  ('prepend', prepend),
  ('remove', remove),
  ('reverse', reverse),
  ('sort', sort),
  ('tail', tail),
  ('take', take),
  ('zip', zip),
]);
